### PR TITLE
Refactor code

### PR DESCRIPTION
**Pull request by ghost**
- [x] Add block keeper for syncing the old blocks
- [x] Handle re-orged blocks
- [x] Refactor code

I tested the deposit and withdraw transaction successfully on my local
Screenshot 2024-08-06 at 14 42 38

Explorer URL: https://sepolia.etherscan.io/tx/0x9ee609a4919f6d9ac3ce673cac0f2f14776acb29c3eabe5fd169f458b9724f20

## How to test my PR?
- Before you start, please make the `docker-compose` first to start the `redis` instance(redis is used for storing the block synced metadata)

```shell
docker-compose up -d
```

- After that, make your `.env` file
```
## thanos notification env
export NETWORK=sepolia

L1_WS_RPC=wss://ethereum-sepolia-rpc.publicnode.com
L2_WS_RPC=wss://rpc.thanos-sepolia.tokamak.network

L1_STANDARD_BRIDGE=0x385076516318551d566CAaE5EC59c23fe09cbF65
L2_STANDARD_BRIDGE=0x4200000000000000000000000000000000000010

L1_USDC_BRIDGE=0xE390EE020Afb7F8e4A2Dc44a71088db2acd72CF3
L2_USDC_BRIDGE=0x4200000000000000000000000000000000000775

SLACK_URL=<your_slack_hook_url>

L1_EXPLORER_URL=https://sepolia.etherscan.io
L2_EXPLORER_URL=https://explorer.thanos-sepolia.tokamak.network

L1_TOKEN_ADDRESSES=0xa30fe40285B8f5c0457DbC3B7C8A280373c40044,0xFF3Ef745D9878AfE5934Ff0b130868AFDDbc58e8,0x42d3b260c761cD5da022dB56Fe2F89c4A909b04A,0x3D752DE8Ec40B655F21D11b02Fa7438AE18687Fc,0x8c4C0FC89382F96E435527D39C9Ec69dDed34E77,0xf8474C2a90B9035e0b431e1789fE76F54d4ce708,0xfFf9976782d46CC05630D1f6eBAb18b2324d6B14
L2_TOKEN_ADDRESSES=0x4200000000000000000000000000000000000486,0xDeadDeAddeAddEAddeadDEaDDEAdDeaDDeAD0000

REDIS_ADDRESS=localhost:6379
```

- Run the application
```
go run cmd/app/main.go listener 
```

- Enjoy with making some deposit/withdrawal transactions with @tokamak-network/thanos-sdk

### CODE REVIEW

**Comment by nguyenzung** [comment]
It is an impressive implementation!

I think there are some things we need to check more. Correct me if I am wrong
- There is a gaps time between syncOldBlock and subscribeNewHead, so it is possible that missing some blocks in this period. 
- In Enqueue(), it is better if not removing oldest blocks. we should remove blocks that need to be removed when reorg occurs. The current impl can lead to a case that a parent hash can not be found in line "if bk.q.Contains(block.Hash().String())"
- I remembered that in each toppic, we have a param is called **Removed**. Maybe we can use this param to detect removed log due to reorg

**Comment by nguyenzung** [comment]
It is an impressive implementation!

I think there are some things we need to check more. Correct me if I am wrong
- There is a gaps time between syncOldBlock and subscribeNewHead, so it is possible that missing some blocks in this period. 
- In Enqueue(), it is better if not removing oldest blocks. we should remove blocks that need to be removed when reorg occurs. The current impl can lead to a case that a parent hash can not be found in line "if bk.q.Contains(block.Hash().String())"
- I remembered that in each toppic, we have a param is called **Removed**. Maybe we can use this param to detect removed log due to reorg

**Comment by 0x6e616d** [comment]
> It is an impressive implementation!
> 
> I think there are some things we need to check more. Correct me if I am wrong
> 
> * There is a gaps time between syncOldBlock and subscribeNewHead, so it is possible that missing some blocks in this period.
> * In Enqueue(), it is better if not removing oldest blocks. we should remove blocks that need to be removed when reorg occurs. The current impl can lead to a case that a parent hash can not be found in line "if bk.q.Contains(block.Hash().String())"
> * I remembered that in each toppic, we have a param is called **Removed**. Maybe we can use this param to detect removed log due to reorg

**Comment by 0x6e616d** [comment]
@nguyenzung 
1. there is a gap if the time to fetch the old blocks is too long(over one block time). For example, the old block is 1, but the current block from `newHead` is 3.
-> It can be lacking with block `2`. But we check the reorg blocks every time we receive one block. That means we can consider this case as the re-orged blocks. -> On the `handleReorgBlocks` function, the parentHash of block `3` is different from the current head(block `1`) -> so this code will fetch the block `2` information -> It can be done.

You can test this case by putting `time.Sleep` on the first line of `subscribeNewHead` function.

2. `circular_queue` is the circular queue that stores the 64 latest blocks. And the parentHash of the next node is the hash of the previous node.
-> In the case of re-orged, we find the re-orged blocks, after that, we will loop the reorged blocks and override the new head.
For example:
```
1. Current circular queue:
1-2-3-4-5

(The correct chain is 1-2-3-4'-5')
2. We receive block 6' -> the parent hash isn't block hash of 5 -> reorg happened
3. We loop to find the canonical chain -> fetch block 5' with hash is the parentHash of 6
4. Until the parentHash of 4' is the hash of 3. We finish.
5. Order them
6. Loop 4' -> 5' -> 6', each block we set head to the circular queue to 1->2-3-4'-5'-6' 
7. We receive block 7' with the parentHash of 7' as the hash of 6' -> continue this process.
```

-> So we finish the process of finding the re-orged blocks, we will have the correct queue.

3. Yes, I missed it, let me update it. We can use `log.Remove` as true or false to know if this event has been removed or added.

File: internal/pkg/listener/service.go [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+		if err != nil {
+			s.l.Errorw("Failed to handle block", "err", err, "block", block)
+			return err
+		}
+		err = s.blockKeeper.SetHead(ctx, block.Header)
+		if err != nil {
+			s.l.Errorw("Failed to set head on the keeper", "err", err, "block", block)
+			return err
+		}
+	}
```
Comment (0x6e616d) [inline internal/pkg/listener/service.go:260]: @nguyenzung ref to 2.6

**Review by ghost — COMMENTED** [review]
@nguyenzung 
1. there is a gap if the time to fetch the old blocks is too long(over one block time). For example, the old block is 1, but the current block from `newHead` is 3.
-> It can be lacking with block `2`. But we check the reorg blocks every time we receive one block. That means we can consider this case as the re-orged blocks. -> On the `handleReorgBlocks` function, the parentHash of block `3` is different from the current head(block `1`) -> so this code will fetch the block `2` information -> It can be done.

You can test this case by putting `time.Sleep` on the first line of `subscribeNewHead` function.

2. `circular_queue` is the circular queue that stores the 64 latest blocks. And the parentHash of the next node is the hash of the previous node.
-> In the case of re-orged, we find the re-orged blocks, after that, we will loop the reorged blocks and override the new head.
For example:
```
1. Current circular queue:
1-2-3-4-5

(The correct chain is 1-2-3-4'-5')
2. We receive block 6' -> the parent hash isn't block hash of 5 -> reorg happened
3. We loop to find the canonical chain -> fetch block 5' with hash is the parentHash of 6
4. Until the parentHash of 4' is the hash of 3. We finish.
5. Order them
6. Loop 4' -> 5' -> 6', each block we set head to the circular queue to 1->2-3-4'-5'-6' 
7. We receive block 7' with the parentHash of 7' as the hash of 6' -> continue this process.
```

-> So we finish the process of finding the re-orged blocks, we will have the correct queue.

3. Yes, I missed it, let me update it. We can use `log.Remove` as true or false to know if this event has been removed or added.

**Comment by 0x6e616d** [comment]
@nguyenzung Oh I got your point 2. Let me think about it. You're correct. We should remove the reorged block on the queue not the first element. Good point

File: go.work.sum
Comment (rlgns98kr) [inline go.work.sum:1]: Can you delete this file ? 🙇🏼‍♂️

Comment (0x6e616d) [reply go.work.sum:1]: Fixed in the commit(https://github.com/tokamak-network/tokamak-thanos-event-listener/pull/11/commits/bd527db8a6c1d93270cae7ae2c3ed109d733060a)

File: internal/pkg/listener/service.go [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+		key := serializeEventRequestWithAddressAndABI(l.Address, l.Topics[0])
+		request := s.RequestByKey(key)
+
+		if request == nil {
+			continue
+		}
+
+		if l.Removed {
+			continue
+		}
```
Comment (0x6e616d) [inline internal/pkg/listener/service.go:281]: Ref to 3.

File: internal/pkg/queue/circular_queue.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	q.data[q.rear] = value
+	q.uniqueData[value] = true
+}
+
+// RemoveAndEnqueue removes and adds an element to the end of the queue.
+// If the queue is full, it overwrites the oldest element.
+func (q *CircularQueue[T]) RemoveAndEnqueue(value T, removed T) {
+	q.Remove(removed)
+	q.Enqueue(value)
+}
```
Comment (0x6e616d) [inline internal/pkg/queue/circular_queue.go:46]: Ref to 2.

File: internal/pkg/repository/block_keeper.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+			return nil, err
+		}
+
+		for _, block := range blocks {
+			keeper.enqueue(block.Header.Hash(), block.Header.Number.Uint64())
+		}
+	}
+
+	keeper.enqueue(common.HexToHash(currentBlockHash), blockNo)
+
```
Comment (0x6e616d) [inline internal/pkg/repository/block_keeper.go:103]: Ref to 2.

**Review by ghost — COMMENTED** [review]
@nguyenzung many thanks @nguyenzung for your helpful comments
1. The gap between the time we run `syncOldBlocks` function and `subscribeNewHead` can be fixed by the function `handleReorgedBlocks` by checking whether the parentHash exists on the queue or not.

2. I updated the queue by adding the new remove function to remove the element by the value. The purpose of this function is to remove the re-orged block hashes if exist not the first element on the queue.

Also, I initiated the block-keeper to get the latest 64 blocks(2 epochs) when starting the event listener service. The purpose of it is to check the re-org of blocks when starting the event listener from scratch or continue from the starting point.

3. I updated the code to ignore the `Removed` log when true

**Comment by nguyenzung** [comment]
It looks great, Nam!

I think maybe we should have other ticket for UnitTest to make sure everything works well

**Review by nguyenzung — APPROVED** [review]

**Pull request merged**